

-----------
diff --git a/tools/libs/light/libxl_create.c b/tools/libs/light/libxl_create.c
index 8856755..0f63359 100644
--- a/tools/libs/light/libxl_create.c
+++ b/tools/libs/light/libxl_create.c
@@ -869,6 +869,9 @@ retry_transaction:
     libxl__xs_mknod(gc, t,
                     GCSPRINTF("%s/data", dom_path),
                     rwperm, ARRAY_SIZE(rwperm));
+    libxl__xs_mknod(gc, t,
+                    GCSPRINTF("%s/error", dom_path),
+                    rwperm, ARRAY_SIZE(rwperm));
     libxl__xs_mknod(gc, t,
                     GCSPRINTF("%s/drivers", dom_path),
                     rwperm, ARRAY_SIZE(rwperm));
-----------------
